src/libs/data-utils.js

```diff
@@ -340,3 +355,249 @@ export const EntityUploader = class EntityUploader extends Component {
+      onSuccess()
+    } catch (e) {
+      if (e.status === 409) {
+        setTakenName(newName)
+        setIsBusy(false)
+      } else {
+        onDismiss()
+        reportError('Unable to rename entity', e)
+      }
+    }
+  }
+
+  return h(Modal, {
+    onDismiss,
+    title: `Rename ${entityName}`,
+    okButton: h(ButtonPrimary, {
+      onClick: doRename,
+      disabled: entityName === newName || errors,
+      tooltip: entityName === newName ? 'No change to save' : Utils.summarizeErrors(errors)
+    }, ['Rename'])
+  }, [
+    h(IdContainer, [id => h(Fragment, [
+      h(FormLabel, { htmlFor: id }, ['New entity name']),
+      h(ValidatedInput, {
+        inputProps: {
+          id,
+          autoFocus: true,
+          placeholder: 'Enter a name',
+          value: newName,
+          onChange: setNewName
+        },
+        error: Utils.summarizeErrors(errors)
+      })
+    ])]),
+    isBusy && spinnerOverlay
+  ])
+}
+
+export const EntityEditor = ({ entityType, entityName, attributeName, attributeValue, entityTypes, workspaceId: { namespace, name }, onDismiss, onSuccess }) => {
+  const initialIsReference = _.isObject(attributeValue) && (attributeValue.entityType || attributeValue.itemsType === 'EntityReference')
+  const initialIsList = _.isObject(attributeValue) && attributeValue.items
+  const initialType = Utils.cond(
+    [initialIsReference, 'reference'],
+    [(initialIsList ? attributeValue.items[0] : attributeValue) === undefined, 'string'],
+    [initialIsList, () => typeof attributeValue.items[0]],
+    typeof attributeValue
+  )
+
+  const [newValue, setNewValue] = useState(() => Utils.cond(
+    [initialIsReference && initialIsList, () => _.map('entityName', attributeValue.items)],
+    [initialIsList, () => attributeValue.items],
+    [initialIsReference, () => attributeValue.entityName],
+    attributeValue
+  ))
+  const [linkedEntityType, setLinkedEntityType] = useState(() => Utils.cond(
+    [initialIsReference && initialIsList, () => attributeValue.items[0] ? attributeValue.items[0].entityType : undefined],
+    [initialIsReference, () => attributeValue.entityType],
+    entityTypes[0]
+  ))
+  const [editType, setEditType] = useState(() => Utils.cond(
+    [initialIsReference, 'reference'],
+    [(initialIsList ? attributeValue.items[0] : attributeValue) === undefined, 'string'],
+    [initialIsList, () => typeof attributeValue.items[0]],
+    typeof attributeValue
+  ))
+  const [isBusy, setIsBusy] = useState()
+  const [consideringDelete, setConsideringDelete] = useState()
+
+  const isList = _.isArray(newValue)
+  const isUnchanged = (initialType === editType && initialIsList === isList && _.isEqual(attributeValue, newValue))
+
+  const doEdit = async () => {
+    try {
+      setIsBusy(true)
+      const prepFn = Utils.switchCase(editType,
+        ['reference', () => v => ({ entityName: _.trim(v), entityType: linkedEntityType })],
+        ['string', () => _.trim]
```

Comment:
Numbers typed into the text box are still stored as strings, and without the conversion here they would end up being saved as strings. We should either convert them as they're entered, or here.